### PR TITLE
perf(decode): 8-slot software prefetch pipeline for sequence execution (#208)

### LINKED ISSUE #208: perf(decode): match prefetch 4-stage pipeline for long-distance matches
## Context

`execute_sequences` reads from history at `cursor - offset` for each match.
For long-distance matches (offset > L2/L3 size), the source line is cold
in cache and the read stalls the pipeline.

Donor `ZSTD_decompressSequencesLong` (huf_decompress.c-adjacent path) uses
a 4-stage software pipeline: decode N+3 ahead of execute N, prefetching the
match source line 3-4 iterations in advance.

## Goal

Port the 4-stage prefetch pipeline:

1. Decode sequence N+3 → emit prefetch for source line of match N+3
2. Decode sequence N+2 → continue prefetching
3. Decode sequence N+1 → source line for N now warm, execute match N
4. Steady state: every iteration prefetches one ahead, executes one behind

## Acceptance criteria

- Pipeline triggered when match offset > prefetch threshold (~32 KiB or
  L2 line residency proxy)
- No regression on small-offset matches (short-offset path stays
  prefetch-free since data is already warm)
- Bench on at least one large-window corpus (`decodecorpus-large-window`
  or equivalent) shows measurable improvement
- All existing decode tests pass; no ratio regression

## Files involved

- zstd/src/decoding/sequence_execution.rs (or wherever execute_sequences
  lives post-#133 interleave)
- zstd/src/decoding/decode_buffer.rs

## References

- Donor: long-distance match prefetch in `ZSTD_decompressSequencesLong`

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Port the donor `ZSTD_decompressSequencesLong_body` shape to the fused decode+execute path: **8-slot software pipeline** (donor `STORED_SEQS = 8`) that decodes ADVANCE sequences ahead of execute, prefetching each match source line at decode time. By the time `buffer.repeat()` reaches the actual load for seq[i], the prefetch issued ADVANCE iterations earlier has had time to pull the line into L1 — hiding DRAM latency on long-distance matches whose source is beyond cache residency.

Closes #208.

### `zstd/src/decoding/decode_buffer.rs`

- New `prefetch_lookahead_match_source(start_idx)` — issues the prefetch hint at a logical buffer position, bound-checks against `buffer.len()` so out-of-range indices (intra-block self-overlap, `wrapping_sub` underflow on a malformed sequence, dictionary-sourced matches predating the current frame) silently drop the hint.
- Wrap-boundary fallback: when the match source straddles the ring's `as_slices()` boundary, prefetches the s1 tail and tops up from s2[0..] so the donor's "up to two cache lines per match" intent doesn't collapse to one (or zero) at the wrap. Short-tail (< 64 B) cases fall back to `prefetch_first_line_l1` so the cache line containing `start_idx` is always hinted.
- Routed through `prefetch_slice` (`_MM_HINT_T0` / `pldl1keep` → L1) matching donor `PREFETCH_L1`.

### `zstd/src/decoding/sequence_section_decoder.rs`

- 8-slot software pipeline: prefill ADVANCE → steady (decode N, prefetch using a `shadow_hist` mirror of `offset_hist`, execute N−ADVANCE) → drain ADVANCE.
- Ring stores raw `[Sequence; 8]`. `do_offset_history` runs both on the local `shadow_hist` during decode-ahead (for exact, not estimated, repcode resolution in the prefetch path) AND on the real `offset_hist` in `execute_one_sequence`. The real `offset_hist` is rolled back from `saved_offset_hist` if the post-loop bitstream check rejects the block, preserving the legacy "partial output, no rewound history" rollback contract on errors.
- **Donor `ZSTD_getOffsetInfo` parity gate**: pipeline only engages when `num_sequences >= 2 * ADVANCE` AND the FSE offsets table shape shows enough long-offset codes — donor's `minShare` (7/256 on 64-bit, 20/256 on 32-bit). The share is computed once when the offsets table is rebuilt (FSE / Predefined modes in `maybe_update_fse_tables`) and cached in `FSEScratch::offsets_long_share`; Repeat-mode blocks reuse the cached value without re-walking the table.
- Compile-time guard on `ADVANCE.is_power_of_two()` so the `i & ADVANCE_MASK` ring indexing stays equivalent to `i % ADVANCE` if `ADVANCE` is ever changed.

### `zstd/src/decoding/prefetch.rs`

- New `prefetch_first_line_l1` helper that issues exactly one L1 prefetch at the first byte of a slice regardless of slice length. `prefetch_slice` no-ops on slices < CACHE_LINE; that's right for bulk prefetch but wrong for the wrap-boundary case where the cache line containing `start_idx` is still the line we need warmed. x86_64 / x86 / aarch64 implementations.

## Test plan

- [x] full nextest suite — 576 passed, 3 skipped
- [x] lint+typecheck clean
- [x] Bench on Linux server (i9-9900K, BMI2/AVX2) on level_3 + level_19 corpora — mixed results, bench variance dominates over the small expected gains on this hardware
- [ ] Flamegraph on large-window corpus to confirm DRAM stall reduction

## Summary by CodeRabbit

* **Performance**
  * Pipelined sequence execution for large blocks to boost decompression throughput
  * Lookahead prefetching and L1-first-line prefetching to reduce memory latency and improve cache efficiency
* **New Features**
  * Tunable pipeline gating based on offset-table characteristics for steadier hot-path execution
* **Tests**
  * Unit tests validating lookahead prefetch behavior and offset-share computation
* **Documentation**
  * Clarified offset-history and pipeline-gate semantics in decoder docs

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/227?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by polaz** [comment]
@coderabbitai review

File: zstd/src/decoding/sequence_section_decoder.rs [hunk 1]
```diff
@@ -141,33 +147,224 @@ pub fn decode_and_execute_sequences<B: super::buffer_backend::BufferBackend>(
...
+            (3, false) => hist[2],
+            // Mirror `do_offset_history_repcode`'s `wrapping_sub(1)` —
+            // not `saturating_sub` — so even on corrupted/0 history
+            // the estimate matches the real resolver. The result is
+            // only consumed by prefetch addressing, where a wild
+            // wrap just lands the hint on an out-of-range slot the
+            // helper drops.
+            (3, true) => hist[0].wrapping_sub(1),
+            _ => 0,
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/sequence_section_decoder.rs]: @coderabbitai Fixed in 01be9195 — replaced estimate_actual_offset with a shadow_hist:[u32;3] advanced via do_offset_history per decode-ahead. The shadow tracks repcode mutations for the ADVANCE-queued sequences, so a fresh huge offset followed by a repcode resolves to the exact post-update value the execute path will later commit — no more pre-update lag suppressing long-distance prefetches.

File: zstd/src/decoding/sequence_section_decoder.rs [hunk 2]
```diff
@@ -141,33 +147,224 @@ pub fn decode_and_execute_sequences<B: super::buffer_backend::BufferBackend>(
...
+                buffer,
+                literals_buffer,
+                &mut lit_cur,
+                literals_buffer_len,
+                offset_hist,
+                exec_seq,
+            )?;
+            seq_sum = seq_sum.wrapping_add(exec_seq.ll).wrapping_add(exec_seq.ml);
+        }
+    } else {
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/sequence_section_decoder.rs]: @coderabbitai Fixed in 01be9195 — added saw_prefetch_needed flag tracked through the prefill loop. If no prefilled sequence resolves above PREFETCH_OFFSET_THRESHOLD (all-short-offset block dominated by warm matches), drain the queued ring sequences in order and continue with the single-pass loop for the rest of the block, skipping ring/steady-state/drain overhead entirely. The pipelined path only stays engaged when at least one prefill sequence actually needs a long-distance prefetch.

File: zstd/src/decoding/sequence_section_decoder.rs [hunk 3]
```diff
@@ -141,32 +154,282 @@ pub fn decode_and_execute_sequences<B: super::buffer_backend::BufferBackend>(
...
+            // at its pre-block value (the success-only commit above
+            // never ran), so restoring from `saved_offset_hist` is
+            // effectively a no-op on the hist side — the explicit
+            // assignment makes the intent unambiguous and protects
+            // against any future refactor that moves the commit
+            // earlier in the pipelined flow.
+            if buffer.try_restore_checkpoint(buffer_checkpoint) {
+                *offset_hist = saved_offset_hist;
+            }
+            return Err(e);
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/sequence_section_decoder.rs:406]: The rollback-on-Err shape here is intentional donor parity, not a divergence.

Donor `ZSTD_decompressSequencesLong_body` (lib/decompress/zstd_decompress_block.c:1832-1990) commits repcodes ONCE at end of block, AFTER `RETURN_ERROR_IF(!BIT_endOfDStream(...))`:

```c
/* save reps for next block */
{ U32 i; for (i=0; i<ZSTD_REP_NUM; i++) dctx->entropy.rep[i] = (U32)(seqState.prevOffset[i]); }
```

If any mid-loop `ZSTD_execSequence` returns an error (`oneSeqSize` is error → `if (ZSTD_isError(oneSeqSize)) return oneSeqSize;`), or the post-loop bitstream-end check fails, that save never runs and `dctx->entropy.rep[i]` stays at its pre-block value. The same applies to `ZSTD_decompressSequences_body` (`zstd_decompress_block.c:1717-1797`) — same end-of-block save pattern, same skip on early return.

Our non-pipelined `execute_one_sequence` currently mutates real `offset_hist` per sequence (an older pattern from the pre-fused-executor era). That's the path that diverges from donor — not the pipelined path I just landed. Bringing the non-pipelined path to donor parity is in scope of a separate cleanup, not this PR.

On the `try_restore_checkpoint` cap-changed edge case: the conditional `if buffer.try_restore_checkpoint(...) { *offset_hist = saved_offset_hist; }` is the same shape used by the existing post-loop bitstream-validation rollback (lines 426-428, pre-existing). Both branches share the "only restore hist when buffer rollback actually succeeded" rule so the hist+buffer pair stays in lockstep — restoring hist while leaving buffer partial would be the actual inconsistency. The cap-changed case implies the frame is already malformed past `MAX_BLOCK_SIZE`; the partial data left in the buffer is discarded by the propagated `Err` anyway.
